arpd.8
.\"
.\" Copyright (c) 2001 Dug Song <dugsong@monkey.org>
.\"
.Dd August 4, 2001
.Dt ARPD 8
.Sh NAME
.Nm farpd
.Nd ARP reply daemon
.Sh SYNOPSIS
.Nm farpd
.Op Fl d
.Op Fl i Ar interface
.Op Ar net ...
.Sh DESCRIPTION
.Nm
replies to any ARP request for an IP address matching the specified
destination
.Ar net
with the hardware MAC address of the specified
.Ar interface ,
but only after determining if another host already claims it.
.Pp
Any IP address claimed by
.Nm
is eventually forgotten after a period of inactivity or after a
hard timeout, and is relinquished if the real owner shows up.
.Pp
This enables a single host to claim all unassigned addresses on a
LAN for network monitoring or simulation.
.Pp
.Nm
exits on an interrupt or termination signal.
.Pp
Note: The program name
.Nm
has been changed in Debian GNU/Linux from the original name
(\fBarpd\fR) to avoid name clash with other ARP daemons.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl d
Do not daemonize, and enable verbose debugging messages.
.It Fl i Ar interface
Listen on
.Ar interface .
If unspecified, 
.Nm 
searches the system interface list for the lowest numbered, configured
``up'' interface (excluding loopback).
.It Ar net
The IP address or network (specified in CIDR notation) or IP address
ranges to claim
(e.g. ``10.0.0.3'', ``10.0.0.0/16'' or ``10.0.0.5-10.0.0.15''). If unspecified,
.Nm
will attempt to claim any IP address it sees an ARP request for.
Mutiple addresses may be specified.
.El
.Sh FILES
.Bl -tag -width /var/run/arpd.pid
.It Pa /var/run/farpd.pid
.El
.Sh SEE ALSO
.Xr pcapd 8 ,
.Xr synackd 8
.Sh BUGS
.Nm
will respond too slowly to ARP requests for some applications. In 
order to ensure that it does not claim existing IP addresses it will send two
ARP request and wait for a reply. This slowness affects the \fBnmap\fR network
scanning tool, and possibly others, which uses by default ARP when scanning
local networks. The answers from 
.Nm
will come after the tool has timeout waiting for the ARP replies and,
consequently, IP addresses claimed by 
.Nm 
will not be discovered.
.Pp
Additionally, 
.Nm
sends the ARP replies to the broadcast address of the network and not to the
host that send the ARP request. Some systems and applications (notably
\fBnmap\fR) will not handled these requests and expect directed ARP replies
(i.e. targeted specifically to the host that sent the request and not to the
network)
.Sh AUTHORS
Dug Song
.Aq dugsong@monkey.org ,
Niels Provos
.Aq provos@citi.umich.edu


debian/changelog
farpd (0.2-11) unstable; urgency=low

  * configure.in: Fix FTBFS: configure: error: libevent not found with the
    patch provided by Sebastian Ramacher (Closes: #676066)
  * debian/rules: Use DEB_BUILD_MULTIARCH instead of DEB_BUILD_GNU_TYPE so
    that libevent.a is found in the correct place.
  * debian/control: Update maintainer's name and email
  * debian/farpd.init: Add dependency on $remote_fs (lintian error)
  * debian/compat: Use debhelper compatibility version 7

 -- Javier Fernández-Sanguino Peña <jfs@debian.org>  Mon, 18 Jun 2012 19:04:19 +0200

farpd (0.2-10) unstable; urgency=low

  * Remove code using event_gotsig and event_sigcb, this external integers
    are deprecated and no longer defined in newer versions of libevent 
    (Closes: #559012)
  * Fix LSB header, farpd requires $network and $syslog

 -- Javier Fernandez-Sanguino Pen~a <jfs@debian.org>  Thu, 10 Dec 2009 02:38:17 +0100

farpd (0.2-9) unstable; urgency=low

  * Describe in the manpage known issues with nmap as described in Honeyd's
    forums. More specifically at
    http://honeyd.org/phpBB2/viewtopic.php?t=473&sid=8be0cc5c1edeb38458d41c15df04bbc0

 -- Javier Fernandez-Sanguino Pen~a <jfs@debian.org>  Thu, 15 May 2008 00:15:59 +0200

farpd (0.2-8) unstable; urgency=low

  * Provide an LSB-complaint init.d script even if disabled by default
    (Closes: #384320)

 -- Javier Fernandez-Sanguino Pen~a <jfs@computer.org>  Fri,  1 Dec 2006 02:15:47 +0100

farpd (0.2-7) unstable; urgency=low

  * Recompile against latest libevent-dev (1.0b) and libdumbnet-dev (1.8)
    (Closes: #294398)
  * Build-Depend against libpcap0.8-dev | libpcap-dev

 -- Javier Fernandez-Sanguino Pen~a <jfs@computer.org>  Tue, 18 Jan 2005 16:50:55 +0100

farpd (0.2-6) unstable; urgency=low

  * Lintian fixes:
    - Updated Standard-Versions
    - Changelog is now UTF-8
  * Fix FTBFS in gcc 3.4 with patch from Andreas Jochens  (Closes: #258857)

 -- Javier Fernandez-Sanguino Pen~a <jfs@computer.org>  Thu,  5 Aug 2004 20:15:16 +0200

farpd (0.2-5) unstable; urgency=low

  * "The day after the tragedy" release
    Esta versión está dedicada a todos aquellos que perdieron
    sus vidas o fueron heridos en el brutal atentado terrorista
    en Madrid de ayer. Mis más sinceras condolencias a sus familias
    y amigos en este triste día.
   (This release is dedicated to all those who lost their lives
    or were injured in the brutal terrorist attack in Madrid
    yesterday. My deepest condolences to their families and friends on
    this sorrowful day.)
  * Arpd name changed in manpage to farpd and changed PID file location
  from /var/run/arpd.pid to /var/run/farpd.pid (Closes #237488)

 -- Javier Fernandez-Sanguino Pen~a <jfs@computer.org>  Fri, 12 Mar 2004 08:43:27 +0100

farpd (0.2-4) unstable; urgency=low

  * Recompiled with latest versions of libdumbevent and libevent 
    (Closes: #219665)

 -- Javier Fernandez-Sanguino Pen~a <jfs@computer.org>  Sun,  9 Nov 2003 17:07:36 +0100

farpd (0.2-3) unstable; urgency=low

  * This is the "I will repeat 100 times: 'I should ldd my binaries
    before uploading'" release
  * Added libevent-dev to Build-Depends (Closes: #200819)

 -- Javier Fernandez-Sanguino Pen~a <jfs@computer.org>  Fri, 11 Jul 2003 18:39:09 +0200

farpd (0.2-2) unstable; urgency=low

  * Added  libdumbnet-dev and  libpcap-dev to build-depends (Closes: #199938)
  * Fixed the configure.in script so that it calls dumbnet instead of
    dnet.

 -- Javier Fernandez-Sanguino Pen~a <jfs@computer.org>  Fri,  4 Jul 2003 11:01:20 +0200

farpd (0.2-1) unstable; urgency=low

  * Initial Release.
  * Changed upstream name of the package from arpd to farpd 
    (after asking Niels Provos for permission)
  * Changed all references to dnet into dumbnet.

 -- Javier Fernandez-Sanguino Pen~a <jfs@computer.org>  Thu, 27 Mar 2003 10:25:02 +0100



arpd.c
/*
 * arpd.c
 *
 * Copyright (c) 2001, 2002 Dug Song <dugsong@monkey.org>
 * Copyright (c) 2002 Niels Provos <provos@citi.umich.edu>
 *
 * $Id: arpd.c,v 1.16 2003/02/09 04:20:40 provos Exp $
 */

#include <sys/param.h>
#include <sys/types.h>
#include <sys/ioctl.h>
#include <sys/stat.h>

#include <pcap.h>

#include <err.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <unistd.h>

/* XXX - libevent */
#undef timeout_pending
#undef timeout_initialized

#include <event.h>
#include <dnet.h>

#define ARPD_MAX_ACTIVE		600
#define ARPD_MAX_INACTIVE	300

#define PIDFILE			"/var/run/farpd.pid"

static pcap_t			*arpd_pcap;
static arp_t			*arpd_arp;
static eth_t			*arpd_eth;
static struct intf_entry	 arpd_ifent;
static int			 arpd_sig;

static void
usage(void)
{
	fprintf(stderr, "Usage: arpd [-d] [-i interface] [net]\n");
	exit(1);
}

/*
 * Expands several command line arguments into a complete pcap filter string.
 * Deals with normal CIDR notation and IP-IP ranges.
 */

static char *
arpd_expandips(int naddresses, char **addresses)
{
	static char filter[1024];
	char line[1024], *p;
	struct addr dst;

	if (naddresses == 0)
		return (NULL);

	filter[0] = '\0';

	while (naddresses--) {
		/* Get current address */
		p = *addresses++;

		if (filter[0] != '\0') {
			if (strlcat(filter, " or ", sizeof(filter)) >= sizeof(filter))
				errx(1, "%s: too many address for filter", 
				    __func__);
		}

		if (addr_pton(p, &dst) != -1) {

			snprintf(line, sizeof(line), "dst %s%s",
			    dst.addr_bits != 32 ? "net ": "", p);
		} else {
			char *first, *second;
			struct addr astart, aend;
			struct in_addr in;
			ip_addr_t istart, iend;

			second = p;

			first = strsep(&second, "-");
			if (second == NULL)
				errx(1, "%s: Invalid network range: %s",
				    __func__, p);

			line[0] = '\0';
			if (addr_pton(first, &astart) == -1 ||
			    addr_pton(second, &aend) == -1)
				errx(1, "%s: bad addresses %s-%s", __func__,
				    first, second);
#ifdef HAVE_BROKEN_DNET
			if (addr_cmp(&astart, &aend) <= 0)
#else
			if (addr_cmp(&astart, &aend) >= 0)
#endif
			    errx(1, "%s: inverted range %s-%s", __func__,
				first, second);

			/* Completely, IPv4 specific */
			istart = ntohl(astart.addr_ip);
			iend = ntohl(aend.addr_ip);
			while (istart <= iend) {
				char single[32];
				int count = 0, done = 0;
				ip_addr_t tmp;

				do {
					ip_addr_t bit = 1 << count;
					ip_addr_t mask;

					mask = ~(~0 << count);
					tmp = istart | mask;

					if (istart & bit)
						done = 1;

					if (iend < tmp) {
						count--;
						mask = ~(~0 << count);
						tmp = istart | mask;
						break;
					} else if (done)
						break;
					
					count++;
				} while (count < IP_ADDR_BITS);

				if (line[0] != '\0')
					strlcat(line, " or ", sizeof(line));
				in.s_addr = htonl(istart);
				snprintf(single, sizeof(single),
				    "dst net %s/%d",
				    inet_ntoa(in), 32 - count);

				strlcat(line, single, sizeof(line));

				istart = tmp + 1;
			}
		}
		
		if (strlcat(filter, line, sizeof(filter)) >= sizeof(filter))
			errx(1, "%s: too many address for filter", 
			    __func__);
	}

	return (filter);
}

static void
arpd_init(char *dev, int naddresses, char **addresses)
{
	struct bpf_program fcode;
	char filter[1024], ebuf[PCAP_ERRBUF_SIZE], *dst;
	intf_t *intf;
	
	dst = arpd_expandips(naddresses, addresses);

	if ((arpd_arp = arp_open()) == NULL)
		err(1, "arp_open");

	if ((intf = intf_open()) == NULL)
		err(1, "intf_open");

	if (dev == NULL) {
		if ((dev = pcap_lookupdev(ebuf)) == NULL)
			errx(1, "pcap_lookupdev: %s", ebuf);
	}
	arpd_ifent.intf_len = sizeof(arpd_ifent);
	strncpy(arpd_ifent.intf_name, dev, sizeof(arpd_ifent.intf_name) - 1);
	arpd_ifent.intf_name[sizeof(arpd_ifent.intf_name) - 1] = '\0';
	
	if (intf_get(intf, &arpd_ifent) < 0)
		err(1, "intf_get");

	if (arpd_ifent.intf_addr.addr_type != ADDR_TYPE_IP ||
	    arpd_ifent.intf_link_addr.addr_type != ADDR_TYPE_ETH)
		errx(1, "bad interface configuration: not IP or Ethernet");
	arpd_ifent.intf_addr.addr_bits = IP_ADDR_BITS;
	
	snprintf(filter, sizeof(filter), "arp %s%s%s and not ether src %s",
	    dst ? "and (" : "", dst ? dst : "", dst ? ")" : "",
	    addr_ntoa(&arpd_ifent.intf_link_addr));
	
	if ((arpd_pcap = pcap_open_live(dev, 128, 0, 500, ebuf)) == NULL)
		errx(1, "pcap_open_live: %s", ebuf);
	
	if (pcap_compile(arpd_pcap, &fcode, filter, 1, 0) < 0 ||
	    pcap_setfilter(arpd_pcap, &fcode) < 0)
		errx(1, "bad pcap filter: %s", pcap_geterr(arpd_pcap));
#if defined(BSD) && defined(BIOCIMMEDIATE)
	{
		int on = 1;
		if (ioctl(pcap_fileno(arpd_pcap), BIOCIMMEDIATE, &on) < 0)
			err(1, "BIOCIMMEDIATE");
	}
#endif
	if ((arpd_eth = eth_open(dev)) == NULL)
		errx(1, "eth_open");

#ifndef LOG_PERROR
#define LOG_PERROR	0
#endif
	openlog("arpd", LOG_PERROR|LOG_PID|LOG_CONS, LOG_DAEMON);
	syslog(LOG_INFO, "listening on %s: %s", dev, filter);
}

static void
arpd_exit(int status)
{
	eth_close(arpd_eth);
	pcap_close(arpd_pcap);
	arp_close(arpd_arp);
	closelog();
	unlink(PIDFILE);

	exit(status);
}

static void
arpd_send(eth_t *eth, int op,
    struct addr *sha, struct addr *spa,
    struct addr *tha, struct addr *tpa)
{
	u_char pkt[ETH_HDR_LEN + ARP_HDR_LEN + ARP_ETHIP_LEN];

	eth_pack_hdr(pkt, tha->addr_eth, sha->addr_eth, ETH_TYPE_ARP);
	arp_pack_hdr_ethip(pkt + ETH_HDR_LEN, op, sha->addr_eth,
	    spa->addr_ip, tha->addr_eth, tpa->addr_ip);
	
	if (op == ARP_OP_REQUEST) {
		syslog(LOG_DEBUG, "%s: who-has %s tell %s", __func__,
		    addr_ntoa(tpa), addr_ntoa(spa));
	} else if (op == ARP_OP_REPLY) {
		syslog(LOG_INFO, "arp reply %s is-at %s",
		    addr_ntoa(spa), addr_ntoa(sha));
	}
	if (eth_send(eth, pkt, sizeof(pkt)) != sizeof(pkt))
		syslog(LOG_ERR, "couldn't send packet: %m");
}

static int
arpd_lookup(struct addr *addr)
{
	struct arp_entry arpent;
	int error;

	if (addr_cmp(addr, &arpd_ifent.intf_addr) == 0) {
		syslog(LOG_DEBUG, "%s: %s at %s", __func__,
		    addr_ntoa(addr), addr_ntoa(&arpd_ifent.intf_link_addr));
		return (0);
	}
	memcpy(&arpent.arp_pa, addr, sizeof(*addr));
	
	error = arp_get(arpd_arp, &arpent);
	
	if (error == -1) {
		syslog(LOG_DEBUG, "%s: no entry for %s", __func__,
		    addr_ntoa(addr));
	} else {
		syslog(LOG_DEBUG, "%s: %s at %s", __func__,
		    addr_ntoa(addr), addr_ntoa(&arpent.arp_ha));
	}
	return (error);
}

static void
arpd_recv_cb(u_char *u, const struct pcap_pkthdr *pkthdr, const u_char *pkt)
{
	struct arp_hdr *arp;
	struct arp_ethip *ethip;
	struct arp_entry src;
	struct timeval tv;
	struct addr pa;

	if (pkthdr->caplen < ETH_HDR_LEN + ARP_HDR_LEN + ARP_ETHIP_LEN)
		return;

	arp = (struct arp_hdr *)(pkt + ETH_HDR_LEN);	/* XXX */
	ethip = (struct arp_ethip *)(arp + 1);
	
	addr_pack(&src.arp_ha, ADDR_TYPE_ETH, ETH_ADDR_BITS,
	    ETH_ADDR_BROADCAST, ETH_ADDR_LEN);
	addr_pack(&src.arp_pa, ADDR_TYPE_IP, IP_ADDR_BITS,
	    ethip->ar_spa, IP_ADDR_LEN);
	    
	switch (ntohs(arp->ar_op)) {
		
	case ARP_OP_REQUEST:
		addr_pack(&pa, ADDR_TYPE_IP, IP_ADDR_BITS,
		    &ethip->ar_tpa, IP_ADDR_LEN);

			arpd_send(arpd_eth, ARP_OP_REPLY,
			    &arpd_ifent.intf_link_addr, &pa,
			    &src.arp_ha, &src.arp_pa);
		break;
		
	case ARP_OP_REPLY:
		addr_pack(&pa, ADDR_TYPE_IP, IP_ADDR_BITS,
		    &ethip->ar_spa, IP_ADDR_LEN);
		break;
	}
}

static void
arpd_recv(int fd, short type, void *ev)
{
	event_add((struct event *)ev, NULL);
	
	if (pcap_dispatch(arpd_pcap, -1, arpd_recv_cb, NULL) < 0)
		syslog(LOG_ERR, "pcap_dispatch: %s", pcap_geterr(arpd_pcap));
}
 
void
terminate_handler(int sig)
{
	arpd_sig = sig;
}

int
arpd_signal(void)
{
	syslog(LOG_INFO, "exiting on signal %d", arpd_sig);
	arpd_exit(0);
	/* NOTREACHED */
	return (-1);
}

int
main(int argc, char *argv[])
{
	struct event recv_ev;
	char *dev;
	int c, debug;
	FILE *fp;

	dev = NULL;
	debug = 0;
	
	while ((c = getopt(argc, argv, "di:h?")) != -1) {
		switch (c) {
		case 'd':
			debug = 1;
			break;
		case 'i':
			dev = optarg;
			break;
		default:
			usage();
			break;
		}
	}
	argc -= optind;
	argv += optind;

	if (argc == 0)
		arpd_init(dev, 0, NULL);
	else
		arpd_init(dev, argc, argv);
	
	if ((fp = fopen(PIDFILE, "w")) == NULL)
		err(1, "fopen");
	
	if (!debug) {
		setlogmask(LOG_UPTO(LOG_INFO));
		
		if (daemon(1, 0) < 0) {
			unlink(PIDFILE);
			err(1, "daemon");
		}
	}
	fprintf(fp, "%d\n", getpid());
	fclose(fp);
	
	chmod(PIDFILE, 0644);

	event_init();
	
	event_set(&recv_ev, pcap_fileno(arpd_pcap), EV_READ,
	    arpd_recv, &recv_ev);
	event_add(&recv_ev, NULL);
	
	/* Setup signal handler */
	if (signal(SIGINT, terminate_handler) == SIG_ERR) {
		perror("signal");
		return (-1);
	}
	if (signal(SIGTERM, terminate_handler) == SIG_ERR) {
		perror("signal");
		return (-1);
	}
	
	event_dispatch();

	return (0);
}
